Adding Untagged Notes to Sidebar

illusaen opened the pull request:
Fixes #1230 

### Fix
* Adds "Untagged Notes" filter to sidebar.
* Incidentally, found bug where filtering Trash notes and then filtering by a tag results in searching for that tag in Trash notes only. I'm assuming that's not the desired behaviour and corrected it; however I can easily change it back.  **_NOTE_: This bug has been fixed with state refactor.**

### Test
#### Untagged Notes
1. Create note with no tags and a note with tags.
2. Open sidebar.
3. Click on "Untagged Notes".
4. See that only the note with no tags shows up in list.

### Screenshots
<details>
  <summary>Before</summary>
  <img src="https://user-images.githubusercontent.com/13263478/108815033-64f1e880-7579-11eb-8d8d-94037c51ad5f.gif" width="100%"/>
</details>

<details>
  <summary>No untagged notes</summary>
  
![no-untagged](https://user-images.githubusercontent.com/13263478/110157645-0c5af080-7dae-11eb-8073-a80ca9edc187.gif)
</details>

<details>
  <summary>Delete tags</summary>
  
![delete-tags-reset-untagged](https://user-images.githubusercontent.com/13263478/110175727-dcb8e200-7dc7-11eb-9b06-70ab2be4732f.gif)

</details>

<details>
  <summary>Long list of tags</summary>
  
![tag-lots](https://user-images.githubusercontent.com/13263478/110157759-2f85a000-7dae-11eb-9100-5a270ccf4800.gif)
</details>

### Release
`RELEASE-NOTES.md` was updated with:
```
### Enhancements

- Added Untagged Notes filter to sidebar.

### Fixes

- Fixed bug where selecting Trash filter and then filtering on a tag resulted in a filter where only notes with that tag in the Trash were shown.
```

illusaen committed:
Adding untagged notes to sidebar. Fixing bug where going from trash to tag filter was showing only notes with that tag in trash.

dmsnell commented on lib/state/ui/reducer.ts:
```diff
@@ -206,6 +208,20 @@ const openedTag: A.Reducer<T.TagHash | null> = (state = null, action) => {
+    case 'SELECT_TRASH':
+    case 'SHOW_ALL_NOTES':
+    case 'OPEN_TAG':
+    case 'TRASH_TAG':
```
why would trashing a tag close out the untagged view?

illusaen replied:
Good point, removing that!

illusaen replied:
No longer pertinent due to state refactor.

dmsnell commented:
@illusaen this is great! it’s also something that’s long been on our list. nice work submitting code that’s very much in the style of the project.

one of the reasons I have personally not attended to this yet is a small state refactor I was interested in doing. as you noted, we have a few booleans describing mutually-exclusive states, whereas we could very well represent the same information with a single state value that cannot be represented in contradictory ways, as is currently the case with `showTrash`, `showUntagged`, and `openedTag`. This would likely be something akin to this:

```ts
type Collection
  = {type: ‘all-notes’}
  | {type: ‘trash’}
  | {type: ‘tag’, tagName: string}
  | {type: ‘untagged’};

const collection: A.Reducer<Collection> = (state = {type: ‘all-notes’}, action) => {
    switch (action.type) {
        case ‘OPEN_TRASH’:
            return {type: ‘trash’};

        case ‘OPEN_TAG’:
            return {type: ‘tag’, tagName: action.tagName};

        case ‘SHOW_ALL_NOTES’:
            return {type: ‘all-notes’};

        case ‘SHOW_UNTAGGED’:
            return {type: ‘untagged’};

        default:
            return state;
    }
}
```
(please overlook any defects in my code example - I’m rarely posting examples that don’t have errors in them because I’m only really concerned about getting the point across)

with this we never have to worry about “closing” the state, such as by monitoring for `SHOW_UNTAGGED` inside of the `showTrash` reducer or attending to those states in the search middleware.

now it may be that this state refactor could turn into a major job and I don’t know if it would be worth it. this PR looks pretty good as-is, but what are your thoughts about such a refactor and would you be interested in trying to perform such a feat for this task?

illusaen commented:
> @illusaen this is great! it’s also something that’s long been on our list. nice work submitting code that’s very much in the style of the project.

Thanks!

> one of the reasons I have personally not attended to this yet is a small state refactor I was interested in doing...
> now it may be that this state refactor could turn into a major job and I don’t know if it would be worth it. this PR looks pretty good as-is, but what are your thoughts about such a refactor and would you be interested in trying to perform such a feat for this task?

I think that refactor would definitely be worth it because of how many chances there are to accidentally break it as it is right now. However, I think there are probably a lot of small edge cases that need to tested so I'd like to split it off from this PR if possible. WDYT?

dmsnell commented:
> I'd like to split it off from this PR if possible. WDYT?

come on now, you're just speaking my language and trying to butter me up 😉 

my only concern with this, which is why I asked your thoughts, is that I don't want a refactor of existing state to stand in the way of your contribution. if you would like to attempt the state refactor in another PR and then come back to this one after that (which should leave this PR in a fairly terse and clear state) then I'm all for that. what I don't want to do is set goalposts that are unreasonable for someone coming in and volunteering some help to the app. this is a long-requested feature and a feature-parity issue with out other platforms.

we will prioritize review for any PRs you submit and be happy to interact with you the way we'd interact with each other on it.

illusaen commented:
> my only concern with this, which is why I asked your thoughts, is that I don't want a refactor of existing state to stand in the way of your contribution. if you would like to attempt the state refactor in another PR and then come back to this one after that (which should leave this PR in a fairly terse and clear state) then I'm all for that. 

Yes, i think this would be the simplest way to do things. :)

> what I don't want to do is set goalposts that are unreasonable for someone coming in and volunteering some help to the app. this is a long-requested feature and a feature-parity issue with out other platforms.
> 

That's fair! I do have a week and a half left with y'all so hopefully the refactor will go faster than that, haha.

> we will prioritize review for any PRs you submit and be happy to interact with you the way we'd interact with each other on it.

Awesome, thanks much! I've noticed y'all are very on the ball about reviewing PRs hehe.

SylvesterWilmott commented:
Thanks for tackling this @illusaen 

Design-wise there's a few things to mention:

- Since this particular filter is only relevant when the user has at least 1 tag, the untagged notes filter should be hidden if there are 0 tags in the tags list.
- You can find the `untagged` icon [here](https://github.com/Automattic/Simplecons/blob/trunk/SVG/24px/untagged.svg) in the Simplecons repo.
- The untagged notes row should appear below tags in the UI in a separate section. See below for illustration.

![Tags](https://user-images.githubusercontent.com/36532468/109957391-cb46db80-7cdc-11eb-90d5-e9de12fb486b.png)

illusaen committed:
Refactoring to use new collection state.

illusaen committed:
Fixing untagged notes design to include new icon and position. Adding no-notes message for untagged notes.

SylvesterWilmott commented:
Looks good and works as expected. Thanks @illusaen

illusaen committed:
Merge branch 'develop' into untagged-sidebar

illusaen committed:
When deleting last tag and untagged notes is selected, will autoselect all notes.

dmsnell committed:
When deleting last tag when inside the untagged view, switch to all notes view

this patch removes the double-dispatch introduced to jump to the all-notes view
when deleting the last tag. that dispatch introduced the unwanted side-effect
of closing the navigation bar.

in this version of the code we introduce a new data point to the `TRASH_TAG`
action which represents the number of remaining tags in the account _after_
trashing the one being removed. the collection reducer and any other reducer
can then use that count to make decisions on how to update without needing
to introduce update cycles.

illusaen committed:
Merge pull request #2721 from Automattic/add/untagged-notes-jump-to-all-notes

When deleting last tag when inside the untagged view, switch to all notes view

illusaen committed:
Pulled out tag filter and count into a selector.

codebykat reviewed (approved):
This is testing out great for me now! Let's :shipit:

dmsnell commented on lib/state/selectors.ts:
```diff
@@ -54,3 +55,8 @@ export const showTrash: S.Selector<boolean> = ({ ui: { collection } }) =>
+export const numberOfNonEmailTags: S.Selector<number> = ({ data }) =>
+  [...data.tags.values()]
+    .filter((tag) => !isEmailTag(tag.name))
+    .map((tag) => tag.name).length;
```
No need for this map - it’s redundant

illusaen replied:
You are 1000% correct. See https://github.com/Automattic/simplenote-electron/pull/2723

dmsnell replied:
Also the selector may even be overkill. It’s fine as is, but I don’t see us needing to know often how many tags there are

illusaen replied:
That's fair. I was actually going to make it return an array of `TagName`s (thus why I had that map in there) so it could be used elsewhere but then I read the comment 🤷‍♀️
